src/Microsoft.DotNet.Wpf/src/PresentationCore/MS/internal/Text/TextInterface/DWriteLoader.cs

```diff
@@ -0,0 +1,100 @@
+
+            if (dwriteCreateFactory == null)
+                throw new InvalidOperationException();
+
+            _dwriteCreateFactory = dwriteCreateFactory;
+        }
+
+        internal static void UnloadDWrite()
+        {
+            ClearDWriteCreateFactoryFunctionPointer();
+        
+            if (_dwrite != IntPtr.Zero)
+            {
+                if (Kernel32.FreeLibrary(_dwrite) != 0)
+                {
+                    int lastError = Marshal.GetLastPInvokeError();
+                    Marshal.ThrowExceptionForHR(HRESULT_FROM_WIN32(lastError));
+                }
+
+                _dwrite = IntPtr.Zero;
+            }
+        }
+
+        internal static delegate* unmanaged<int, void*, void*, int> GetDWriteCreateFactoryFunctionPointer()
+        {
+            return _dwriteCreateFactory;
+        }
+
+        private static void ClearDWriteCreateFactoryFunctionPointer()
+        {
+            _dwriteCreateFactory = null;
+        }
+
+        private static IntPtr LoadDWriteLibraryAndGetProcAddress(out delegate* unmanaged<int, void*, void*, int> DWriteCreateFactory)
+        {
+            IntPtr hDWriteLibrary = IntPtr.Zero;
+
+            // KB2533623 introduced the LOAD_LIBRARY_SEARCH_SYSTEM32 flag. It also introduced
+            // the AddDllDirectory function. We test for presence of AddDllDirectory as an 
+            // indirect evidence for the support of LOAD_LIBRARY_SEARCH_SYSTEM32 flag. 
+            IntPtr hKernel32 = Kernel32.GetModuleHandleW(Libraries.Kernel32);
+
+            if (hKernel32 != IntPtr.Zero)
+            {
+                if (Kernel32.GetProcAddress(hKernel32, "AddDllDirectory") != IntPtr.Zero)
+                {
+                    // All supported platforms newer than Vista SP2 shipped with dwrite.dll.
+                    // On Vista SP2, the .NET servicing process will ensure that a MSU containing 
+                    // dwrite.dll will be delivered as a prerequisite - effectively guaranteeing that 
+                    // this following call to LoadLibraryEx(dwrite.dll) will succeed, and that it will 
+                    // not be susceptible to typical DLL planting vulnerability vectors.
+                    hDWriteLibrary = Kernel32.LoadLibraryExW("dwrite.dll", IntPtr.Zero, Kernel32.LOAD_LIBRARY_SEARCH_SYSTEM32);
+                }
+                else
+                {
+                    // LOAD_LIBRARY_SEARCH_SYSTEM32 is not supported on this OS. 
+                    // Fall back to using plain ol' LoadLibrary
+                    // There is risk that this call might fail, or that it might be
+                    // susceptible to DLL hijacking. 
+                    hDWriteLibrary = Kernel32.LoadLibraryW("dwrite.dll");
+                }
+            }
+
+            if (hDWriteLibrary != IntPtr.Zero)
+            {
+                DWriteCreateFactory = (delegate* unmanaged<int, void*, void*, int>)Kernel32.GetProcAddress(hDWriteLibrary, "DWriteCreateFactory");
```

Comment:
The `NativeLibrary` class provides decent support of `DllSearchPath`. P/Invoke to `LoadLibrary` and `FreeLibrary` can be removed then.

Reply:
I was always under the impression that `NativeLibrary` should be used when it should work cross platform but since WPF is Windows only, `NativeLibrary` would only add unnecessary overhead instead of direct P/Invoke. (I don't know if this is the official recommendation, this is only from memory so it might be wrong)

Reply:
> but since WPF is Windows only, NativeLibrary would only add unnecessary overhead instead of direct P/Invoke. (I don't know if this is the official recommendation, this is only from memory so it might be wrong)

I would advocate for using `NativeLibrary` and official .NET APIs as opposed to calling OS specific functions. The overhead here is minimal at best and would be a one time cost relative to how often the function pointer itself. Calling the `LoadLibrary()` and `GetProcAddress` directly force a lot of thinking about flags and system state. Using `NativeLibrary` directly avoids any potential confusion and handles all security concerns.

Reply:
Done in 49550d13a0069b9529630e5f37671825b02b1d73